principles/coding-principles.md

```diff
@@ -65,7 +65,17 @@ It’s tempting to build an inheritance tree of objects extending others and red
 This includes estimates of work - if they are to be used for assigning budget or predicting delivery dates, they should always come with a level of confidence, or a best/worst case range (e.g. “between x and y days” or “maybe x days, but i’m only 50% confident of that”)
 The larger the chunk of work you’re estimating, the more inaccurate you will be - so if it doesn’t fit into a single sprint, break it down until it does
 
-## Ruby-specific principles:
+## Object Oriented Design principles:
+
+### 1. Use [SOLID](https://scotch.io/bar-talk/s-o-l-i-d-the-first-five-principles-of-object-oriented-design)
+SOLID helps to create code that is easy to maintain and extend.
+
+### 2. Use [Design Patterns](https://www.dofactory.com/net/design-patterns)
```

Comment:
I like design patterns and agree we should recommend using them, however I've experienced considerable pain at the hands of an ill-deployed pattern. One example is singletons being used to pass mutable global state all over the place which we've probably all seen (or done before we realised!) and which does not play nice with SOLID at all.

I've also quite often seen over-enthusiastic deployment of patterns create highly abstracted code which is hard for developers to work with due to the cognitive load. Often these patterns did not add anything meaningful in terms of efficiency or reliability in return for degrading maintainability. Sometimes they had a significant negative effect on performance, introduced new bugs or memory leaks etc.

So I think we need to weigh up a couple of things in this guidance
- Recommending the sensible and wise use of software patterns
- Discouraging the use of patterns at the expense of maintainability, legibility or performance (i.e. having a good reason for using them)